apollo-router/src/drop_watch.rs

```diff
@@ -0,0 +1,141 @@
+//! Provide a [`DropWatch`] utility.
+//!
+//! Accept a FnOnce and carefully watch over it until we are dropped.
+//!
+//! Useful when we want to join threads which will ultimately be terminated
+//! within an asynchronous context.
+
+use std::mem::ManuallyDrop;
+use std::sync::atomic::AtomicBool;
+use std::sync::atomic::Ordering;
+use std::sync::Arc;
+use std::thread::JoinHandle;
+
+pub(crate) struct DropWatch {
+    watcher_handle: ManuallyDrop<JoinHandle<()>>,
```

Comment:
That's what I had originally, but the ManuallyDrop approach is cleaner. Like I said yesterday, `take` is only marked as `unsafe` because it's unsafe to do unless you know what you are doing. It's the best way to do it from a `Drop`, so that's what the comment and link to the documentation are recording in the `Drop` implementation.